prow/cmd/deck/tide.go

```diff
@@ -147,68 +151,118 @@ func (ta *tideAgent) updateHistory() error {
+		// If the orgrepo is associated with no tenantID OR the default tenantID we ignore it here.
+		// This prevents already IDd pools from getting the default ID assigned to them when their orgrepo is not associated with an OrgRepo.
+		// Pools with no tenantID and with default tenantID behave the same, so adding the default ID just causes issues
+		if orgRepoID != "" && orgRepoID != config.DefaultTenantID {
+			curIDs.Insert(orgRepoID)
+		}
+		if len(ta.tenantIDs) > 0 {
+			// Deck has tenantIDs
+			if ta.matchingIDs(curIDs.List()) {
+				// tenantIDs match with the pool
+				filtered = append(filtered, pool)
+			}
+		} else if needsHide := matches(pool.Org+"/"+pool.Repo, ta.hiddenRepos()); needsHide {
+			// Deck has no tenantIDs, and the Pool needs to be hidden
+			if ta.showHidden || ta.hiddenOnly {
+				// Show hidden or hidden only is true
+				filtered = append(filtered, pool)
+			}
+		} else if !ta.hiddenOnly && noTenantIDOrDefaultTenantID(curIDs.List()) {
+			// Pool is not hidden and has no tenantID and Deck is not hidden only.
 			filtered = append(filtered, pool)
 		}
 	}
 	return filtered
 }
 
-func (ta *tideAgent) filterHiddenHistory(hist map[string][]history.Record) map[string][]history.Record {
-	if len(ta.hiddenRepos()) == 0 {
-		return hist
+func noTenantIDOrDefaultTenantID(ids []string) bool {
+	for _, id := range ids {
+		if id != "" && id != config.DefaultTenantID {
+			return false
+		}
+	}
+	return true
+}
+
+func recordIDs(records []history.Record) sets.String {
+	res := sets.String{}
+	for _, record := range records {
+		res.Insert(record.TenantIDs...)
 	}
+	return res
+}
 
+func (ta *tideAgent) filterHistory(hist map[string][]history.Record) map[string][]history.Record {
 	filtered := make(map[string][]history.Record, len(hist))
 	for pool, records := range hist {
-		needsHide := matches(strings.Split(pool, ":")[0], ta.hiddenRepos())
-		if needsHide && ta.showHidden {
-			filtered[pool] = records
-		} else if needsHide == ta.hiddenOnly {
+		orgRepo := strings.Split(pool, ":")[0]
+		curIDs := recordIDs(records).Insert()
+		orgRepoID := ta.cfg.GetProwJobDefault(orgRepo, "*").TenantID
+		// If the orgrepo is associated with no tenantID OR the default tenantID we ignore it here.
+		// This prevents already IDd History from getting the default ID assigned to them when their orgrepo is not associated with an OrgRepo.
+		// History with no tenantID and with default tenantID behave the same, so adding the default ID just causes issues
+		if orgRepoID != "" && orgRepoID != config.DefaultTenantID {
+			curIDs.Insert(orgRepoID)
+		}
+		if len(ta.tenantIDs) > 0 {
+			if ta.matchingIDs(curIDs.List()) {
+				// Deck has tenantIDs and they match with the History
+				filtered[pool] = records
+			}
+		} else if needsHide := matches(orgRepo, ta.hiddenRepos()); needsHide {
+			if ta.showHidden || ta.hiddenOnly {
+				filtered[pool] = records
+			}
+		} else if !ta.hiddenOnly && noTenantIDOrDefaultTenantID(curIDs.List()) {
```

Comment:
Please refactor some of this common code into another shared function. It would be nice to have one function for the decision of whether we should keep an item based on the deck configuration and the hidden status + tenant IDs of the item.

Reply:
Done. Tried thinking about how to do this before but my brain was fried and couldn't figure it out :) Looks much nicer now